src/cashaddr.cpp

```diff
@@ -0,0 +1,304 @@
+            // {2}k(x) = {15}*x^7 + {6}*x^6 + {27}*x^5 + {22}*x^4 + {27}*x^3 +
+            //           {6}*x^2 + {15}*x + {2}
+            c ^= 0x79b76d99e2;
+        }
+
+        if (c0 & 0x04) {
+            // {4}k(x) = {30}*x^7 + {12}*x^6 + {31}*x^5 + {5}*x^4 + {31}*x^3 +
+            //           {12}*x^2 + {30}*x + {4}
+            c ^= 0xf33e5fb3c4;
+        }
+
+        if (c0 & 0x08) {
+            // {8}k(x) = {21}*x^7 + {24}*x^6 + {23}*x^5 + {10}*x^4 + {23}*x^3 +
+            //           {24}*x^2 + {21}*x + {8}
+            c ^= 0xae2eabe2a8;
+        }
+
+        if (c0 & 0x10) {
+            // {16}k(x) = {3}*x^7 + {25}*x^6 + {7}*x^5 + {20}*x^4 + {7}*x^3 +
+            //            {25}*x^2 + {3}*x + {16}
+            c ^= 0x1e4f43e470;
+        }
+    }
+
+    return c;
+}
+
+/**
+ * Convert to lower case.
+ *
+ * Assume the input is a character.
+ */
+inline uint8_t LowerCase(uint8_t c) {
+    // ASCII black magic.
+    return c | 0x20;
+}
+
+/**
+ * Expand the address prefix for the checksum computation.
+ */
+data ExpandPrefix(const std::string &prefix) {
+    data ret;
+    ret.resize(prefix.size() + 1);
+    for (size_t i = 0; i < prefix.size(); ++i) {
+        ret[i] = prefix[i] & 0x1f;
+    }
+
+    ret[prefix.size()] = 0;
+    return ret;
+}
+
+/**
+ * Verify a checksum.
+ */
+bool VerifyChecksum(const std::string &prefix, const data &values) {
+    /**
+     * PolyMod computes what value to xor into the final values to make the
+     * checksum 0. However, if we required that the checksum was 0, it would be
+     * the case that appending a 0 to a valid list of values would result in a
+     * new valid list. For that reason, cashaddr requires the resulting checksum
+     * to be 1 instead.
+     */
+    return PolyMod(Cat(ExpandPrefix(prefix), values)) == 1;
+}
+
+/**
+ * Create a checksum.
+ */
+data CreateChecksum(const std::string &prefix, const data &values) {
+    data enc = Cat(ExpandPrefix(prefix), values);
+    // Append 8 zeroes.
+    enc.resize(enc.size() + 8);
+    // Determine what to XOR into those 8 zeroes.
+    uint64_t mod = PolyMod(enc) ^ 1;
+    data ret(8);
+    for (size_t i = 0; i < 8; ++i) {
+        // Convert the 5-bit groups in mod to checksum values.
+        ret[i] = (mod >> (5 * (7 - i))) & 0x1f;
+    }
+
+    return ret;
+}
+
+} // namespace
+
+namespace cashaddr {
+
+/**
+ * Encode a cashaddr string.
+ */
+std::string Encode(const std::string &prefix, const data &values) {
+    data checksum = CreateChecksum(prefix, values);
+    data combined = Cat(values, checksum);
+    std::string ret = prefix + ':';
+
+    ret.reserve(ret.size() + combined.size());
+    for (uint8_t c : combined) {
+        ret += CHARSET[c];
+    }
+
+    return ret;
+}
+
+/**
+ * Decode a cashaddr string.
+ */
+std::pair<std::string, data> Decode(const std::string &str,
+                                    const std::string &default_prefix) {
+    // Go over the string and do some sanity checks.
```

Comment:
what's the longest cashaddr?  Let's add this at the top:
if (str.size() > X) return {};

That way the rest of the routine is protected from someone sending a very long string...

Reply:
I tried just using a length of 42 + prefixlength + 1 for the separator...it generally works but many of tests then fail.  Will have to research this a bit.